index.d.ts

```diff
@@ -0,0 +1,758 @@
+   * @param {number} i an index
+   * @param {number} j another index
+   * @returns {Array} x
+   */
+  export function swap$<T>(x: T[], i: number, j: number): T[];
+
+  /**
+   * Exchanges two values.
+   * @param {Array} x an array
+   * @param {number} i an index
+   * @param {number} j another index
+   * @returns {Array}
+   */
+  export function swap<T>(x: T[], i: number, j: number): T[];
+
+  /**
+   * Gets values except first.
+   * @param {Array} x an array
+   * @returns {Array} except first
+   */
+  export function tail<T>(x: T[]): T[];
+
+  /**
+   * Gives union of first array with another.
+   * @param {Array} x an array (updated)
+   * @param {Array} y another array
+   * @param {function?} fn compare function (a, b)
+   * @returns {Array} x
+   */
+  export function union$<T>(x: T[], y: T[], fn?: (a: T, b: T) => boolean): T[];
+
+  /**
+   * Gives union of first array with another.
+   * @param {Array} x an array
+   * @param {Array} y another array
+   * @param {function?} fn compare function (a, b)
+   * @returns {Array}
+   */
+  export function union<T>(x: T[], y: T[], fn?: (a: T, b: T) => boolean): T[];
+
+  /**
+   * Gives union of first array with another.
+   * @param {Array} x an array (updated)
+   * @param {Array} y another array
+   * @param {function?} fn map function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array} x
+   */
+  export function unionOn$<T>(x: T[], y: T[], fn?: (v: T, i: number, x: T[]) => T[], ths?: object): T[];
+
+  /**
+   * Gives union of first array with another.
+   * @param {Array} x an array
+   * @param {Array} y another array
+   * @param {function?} fn map function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array}
+   */
+  export function unionOn<T>(x: T[], y: T[], fn?: (v: T, i: number, x: T[]) => T[], ths?: object): T[];
+
+  /**
+   * Removes duplicate elements.
+   * @param {Array} x an array
+   * @param {function?} fn compare function (a, b)
+   * @returns {Array} unique values
+   */
+  export function unique<T>(x: T[], fn?: (a: T, b: T) => boolean): T[];
+
+  /**
+   * Removes duplicate elements.
+   * @param {Array} x an array
+   * @param {function?} fn map function (v, i, x)
+   * @param {object?} ths this argument
+   * @returns {Array} unique values
+   */
+  export function uniqueOn<T>(x: T[], fn?: (v: T, i: number, x: T[]) => T[], ths?: object): T[];
+
+  /**
+   * Adds values to the start.
+   * @param {Array} x an array
+   * @param {...any} vs values to add
+   * @returns {Array} unshifted
+   */
+  export function unshift<T>(x: T[], ...vs: T[]): T[];
+
+  /**
+   * Combines values from n arrays, with a function.
+   * @param {Array<Array>} xs n arrays
+   * @param {function?} fn combine function (a, b, c, ...)
+   * @param {object?} ths this argument
+   * @returns {Array<Array>} combined values
+   */
+  export function zip<T>(xs: T[][], fn?: (...args: T[]) => T, ths?: object): T[] | T[][];
```

Comment:
Lets say return type depends upon what `fn` returns:

```suggestion
  export function zip<T,U>(xs: T[][], fn?: (...args: T[]) => U, ths?: object): U[];
```

Reply:
Noted, thanks